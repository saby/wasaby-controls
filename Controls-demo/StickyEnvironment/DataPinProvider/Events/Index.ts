import { Memory } from 'Types/source';
import { SyntheticEvent } from 'UI/Vdom';
import { IEdgesData } from 'Controls/stickyEnvironment';
import { Control, IControlOptions, TemplateFunction } from 'UI/Base';
import { generateData } from 'Controls-demo/list_new/DemoHelpers/DataCatalog';
import * as template from 'wml!Controls-demo/StickyEnvironment/DataPinProvider/Events/Index';

export default class Index extends Control<IControlOptions> {
    protected _template: TemplateFunction = template;

    protected _topEdgeData: number;
    protected _bottomEdgeData: number;

    protected _viewSource: Memory;
    protected _navigation: object = {
        source: 'page',
        view: 'infinity',
        sourceConfig: {
            page: 10,
            pageSize: 50,
            hasMore: false,
        },
    };

    private _dataArray: { key: number; title: string }[] = generateData<{
        key: number;
        title: string;
    }>({
        count: 1000,
        entityTemplate: { title: 'lorem' },
    });

    protected _beforeMount(): void {
        this._viewSource = new Memory({
            keyProperty: 'key',
            data: this._dataArray,
        });
    }

    protected _onEdgesDataChanged(
        event: SyntheticEvent,
        data: IEdgesData<number>
    ): void {
        this._topEdgeData = data.top.above || data.top.below;
        this._bottomEdgeData = data.bottom.above;
    }
}
