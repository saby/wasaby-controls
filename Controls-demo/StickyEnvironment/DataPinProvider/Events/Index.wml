<div class="controls-demo_Wrapper controlsDemo_fixedWidth500">
   <div data-qa="controls-demo__text_top">Top edge data id: {{_topEdgeData}}</div>
   <div data-qa="controls-demo__text_bottom">Bottom edge data id: {{_bottomEdgeData}}</div>

   <Controls.stickyEnvironment:DataPinProvider on:edgesDataChanged="_onEdgesDataChanged()">
      <Controls.scroll:Container attr:style="height: 500px;">
         <Controls.list:View
            source="{{_viewSource}}"
            keyProperty="key"
            navigation="{{_navigation}}"
         >
            <ws:virtualScrollConfig page="{{50}}" pageSize="{{50}}"/>
            <ws:itemTemplate>
               <ws:partial template="Controls/list:ItemTemplate">
                  <ws:contentTemplate>
                     <Controls.stickyEnvironment:DataPinContainer data="{{itemTemplate.item.contents.get('key')}}">
                        <strong>Запись с id="{{itemTemplate.item.contents.get('key')}}"</strong>
                     </Controls.stickyEnvironment:DataPinContainer>

                     {{itemTemplate.item.contents.get('title')}}
                     <i data-qa="Controls-Demo_StickyEnvironment__hooks">h</i>
                  </ws:contentTemplate>
               </ws:partial>
            </ws:itemTemplate>
         </Controls.list:View>
      </Controls.scroll:Container>
   </Controls.stickyEnvironment:DataPinProvider>
</div>
