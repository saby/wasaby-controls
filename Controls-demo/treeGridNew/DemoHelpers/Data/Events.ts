import { IColumn } from 'Controls/grid';
import { IData } from '../Interface';

export const Events = {
    getColumns: (): IColumn[] => {
        return [
            {
                displayProperty: 'title',
                width: '',
            },
        ];
    },
    getDataBeforeExpanded: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataBeforeCollapsed: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataCollapsedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataExpandedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
};
