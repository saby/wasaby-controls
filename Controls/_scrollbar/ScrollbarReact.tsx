import * as React from 'react';
import { ResizeObserverUtil } from 'Controls/sizeUtils';
import { Container as DnDContainer } from 'Controls/dragnDrop';
import { SyntheticEvent } from 'Vdom/Vdom';
import { detection } from 'Env/Env';
import { createElement, delimitProps } from 'UICore/Jsx';
import { getArgs } from 'UICore/Events';
import { usePreviousScrollbarProps } from './Hooks/UsePreviousScrollbarProps';
import 'css!Controls/scrollbar';

enum ScrollDirection {
    Vertical = 'vertical',
    Horizontal = 'horizontal',
}

/**
 * Тонкий скроллбар.
 * @class Controls/_scrollbar/Scrollbar
 * @extends UI/Base:Control
 * @implements Controls/scrollbar:IScrollbars
 * @public
 * @implements Controls/interface:IControl
 * @demo Controls-demo/Scroll/Scrollbar/Default/Index
 */

interface IThumbProps {
    setThumbRef: Function;
    getThumbWrapperStyle: Function;
    getPaddingClasses: Function;
    thumbMouseDownHandler: React.MouseEventHandler;
    thumbThickness: string;
    direction: string;
    thumbStyle: string;
    dragging: boolean;
}

interface IResizeObserverEntry {
    readonly contentRect: DOMRectReadOnly;
    readonly target: Element;
}

interface IDraggingData {
    dragPointOffset?: number;
    size?: number;
    offset?: number;
}

const MIN_SIZE = 36;

function getMouseCoord(nativeEvent: Event, direction: string): number {
    let offset: number;
    const offsetAxis =
        direction === ScrollDirection.Vertical ? 'pageY' : 'pageX';

    if (nativeEvent instanceof MouseEvent) {
        offset = nativeEvent[offsetAxis];
    } else {
        offset = (nativeEvent as TouchEvent).touches[0][offsetAxis];
    }

    return offset;
}

function calcWheelDelta(delta: number): number {
    if (detection.firefox) {
        const additionalWheelOffset = 100;
        return Math.sign(delta) * additionalWheelOffset;
    }
    return delta;
}

const thumbNode = React.forwardRef((props: IThumbProps, ref) => {
    return (
        <div
            ref={(el: HTMLDivElement) => {
                props.setThumbRef(el);
                ref(el);
            }}
            className={
                `controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-${props.thumbThickness} controls-VScrollbar__thumbWrapper_${props.direction} ` +
                props.getPaddingClasses()
            }
            style={props.getThumbWrapperStyle()}
            onMouseDown={props.thumbMouseDownHandler}
        >
            <div
                className={
                    `controls-VScrollbar__thumb controls-VScrollbar__thumb_${props.thumbStyle} ` +
                    `controls-VScrollbar__thumb_size-${props.thumbThickness} ` +
                    `controls-VScrollbar__thumb_${props.thumbStyle} ` +
                    `controls-VScrollbar__thumb_${props.thumbStyle}_${props.direction} ` +
                    `controls-VScrollbar__thumb_${props.direction} ` +
                    `controls-VScrollbar__thumb_${props.direction}_size-${props.thumbThickness} ` +
                    `${
                        props.dragging
                            ? 'controls-VScrollbar__thumb_dragging_size-' +
                              props.thumbThickness +
                              ' controls-VScrollbar__thumb_dragging_' +
                              props.thumbStyle
                            : ''
                    }`
                }
                data-qa="VScrollbar__thumb"
            ></div>
        </div>
    );
});

function getThumbPosition(
    scrollbarSize: number,
    scrollbarOffset: number,
    mouseCoord: number,
    thumbSize: number,
    thumbSizeCompensation: number
): number {
    let thumbPosition: number;
    // ползунок должен оказываться относительно текущей позииции смещенным
    // при клике на половину своей высоты
    // при перетаскивании на то, расстояние, которое было до курсора в момент начала перетаскивания
    thumbPosition = mouseCoord - scrollbarOffset - thumbSizeCompensation;
    thumbPosition = Math.max(0, thumbPosition);
    thumbPosition = Math.min(thumbPosition, scrollbarSize - thumbSize);
    return thumbPosition;
}

function ScrollbarReact(
    {
        position = 0,
        contentSize,
        direction = 'vertical',
        thumbThickness = 'l',
        thumbStyle = 'accented',
        paddings = { start: false, end: false },
        ...props
    },
    forwardedRef: React.ForwardedRef<unknown>
): React.ReactElement {
    const scrollbarRef = React.useRef(null);
    const thumbRef = React.useRef(null);
    const dndRef = React.useRef(null);
    const [scrollbarSize, setScrollbarSize] = React.useState(undefined);
    const [dragging, setDragging] = React.useState(false);
    const [draggingData, setDraggingData] = React.useState<IDraggingData>({});
    const [thumbPosition, setThumbPosition] = React.useState(0);

    const { userAttrs } = delimitProps(props);

    const [resizeObserver] = React.useState(() => {
        return new ResizeObserverUtil(undefined, resizeHandler.bind(this));
    });

    const prevProps = usePreviousScrollbarProps({ contentSize });

    React.useLayoutEffect(() => {
        resizeObserver.observe(scrollbarRef.current);
        return () => {
            resizeObserver.unobserve(scrollbarRef.current);
        };
    }, []);

    React.useLayoutEffect(() => {
        if (detection.isIE && prevProps?.contentSize !== contentSize) {
            setScrollbarSize(
                direction === 'vertical'
                    ? scrollbarRef.current.clientHeight
                    : scrollbarRef.current.clientWidth
            );
        }
    });

    const thumbSize = React.useMemo(() => {
        if (!thumbRef.current) {
            return;
        }

        const viewportRatio = scrollbarSize / contentSize;
        const thumbSize = scrollbarSize * viewportRatio;
        // MIN_SIZE сделан для тестов, в юнит-тестах нет стилей, заданных на классах из-за этого
        // getComputedStyle возвращает NaN.
        const minSize =
            parseFloat(
                getComputedStyle(thumbRef.current)[
                    direction === ScrollDirection.Vertical
                        ? 'min-height'
                        : 'min-width'
                ]
            ) || MIN_SIZE;
        return Math.max(minSize, thumbSize);
    }, [scrollbarSize, contentSize, thumbRef.current]);

    React.useLayoutEffect(() => {
        const newThumbPosition = getThumbCoordByScroll(
            scrollbarSize,
            thumbSize,
            contentSize,
            position
        );
        setThumbPosition(newThumbPosition);
    }, [contentSize, scrollbarSize, thumbSize, position]);

    function getThumbCoordByScroll(
        scrollbarSize: number,
        thumbSize: number,
        contentSize: number,
        position: number
    ): number {
        if (scrollbarSize === undefined) {
            return 0;
        }

        // ползунок перемещается на расстояние равное высоте скроллбара - высота ползунка
        const availableScale = scrollbarSize - thumbSize;
        // скроллить можно на высоту контента, за вычетом высоты контейнера = высоте скроллбара
        const availableScroll = contentSize - scrollbarSize;
        // решаем пропорцию, известна координата ползунка, высота его перемещения и величину скроллящегося контента
        return (position * availableScale) / availableScroll;
    }

    function getScrollCoordByThumb(
        scrollbarSize: number,
        thumbSize: number,
        contentSize: number,
        position: number
    ): number {
        if (scrollbarSize === undefined) {
            return 0;
        }

        // ползунок перемещается на расстояние равное высоте скроллбара - высота ползунка
        const availableScale = scrollbarSize - thumbSize;
        // скроллить можно на высоту контента, за вычетом высоты контейнера = высоте скроллбара
        const availableScroll = contentSize - scrollbarSize;
        // решаем пропорцию, известна координата ползунка, высота его перемещения и величину скроллящегося контента
        return (position * availableScroll) / availableScale;
    }

    function resizeHandler(entries: IResizeObserverEntry[]): void {
        setScrollbarSize(
            direction === 'vertical'
                ? entries[0].contentRect.height
                : entries[0].contentRect.width
        );
    }

    function getCurrentCoords(direction: string): {
        size: number;
        offset: number;
    } {
        let offsetValue: number;
        let sizeValue: number;

        const scrollBarClientRect =
            scrollbarRef.current.getBoundingClientRect();
        if (direction === ScrollDirection.Vertical) {
            offsetValue = scrollBarClientRect.top;
            sizeValue = scrollBarClientRect.height;
        } else {
            offsetValue = scrollBarClientRect.left;
            sizeValue = scrollBarClientRect.width;
        }
        return {
            offset: offsetValue,
            size: sizeValue,
        };
    }

    function thumbMouseDownHandler(event: React.MouseEvent): void {
        // Отключаем выделение пока идет драг.
        event.preventDefault();
        event.stopPropagation();
        scrollbarBeginDragHandler(event);
    }

    function scrollbarMouseDownHandler(event: React.MouseEvent): void {
        const currentCoords = getCurrentCoords(direction);
        const mouseCoord = getMouseCoord(event.nativeEvent, direction);

        const thumbPosition = getThumbPosition(
            currentCoords.size,
            currentCoords.offset,
            mouseCoord,
            thumbSize,
            thumbSize / 2
        );

        const newPosition = getScrollCoordByThumb(
            currentCoords.size,
            thumbSize,
            contentSize,
            thumbPosition
        );
        setThumbPosition(newPosition);
        props.onPositionchanged(undefined, newPosition);
    }

    function touchStartHandler(event: React.TouchEvent): void {
        if (direction === ScrollDirection.Horizontal) {
            scrollbarBeginDragHandler(event);
        }
    }

    function scrollbarBeginDragHandler(
        event: SyntheticEvent | React.MouseEvent | React.TouchEvent
    ): void {
        const verticalDirection = direction === ScrollDirection.Vertical;
        const thumbOffset =
            thumbRef.current.getBoundingClientRect()[
                verticalDirection ? 'top' : 'left'
            ];
        const mouseCoord = getMouseCoord(event.nativeEvent, direction);

        const currentCoords = getCurrentCoords(direction);
        const dragPointOffset = mouseCoord - thumbOffset;
        setDraggingData({ ...currentCoords, dragPointOffset });
        dndRef.current.startDragNDrop(null, event);
    }

    function scrollbarStartDragHandler(): void {
        setDragging(true);
        if (props.onDraggingchanged) {
            props.onDraggingchanged(undefined, true);
        }
    }

    function scrollbarEndDragHandler(): void {
        if (dragging) {
            setDragging(false);
            if (props.onDraggingchanged) {
                props.onDraggingchanged(undefined, false);
            }
        }
    }

    function scrollbarOnDragHandler(
        event: SyntheticEvent<Event>,
        dragObject: { domEvent: Event }
    ): void {
        const [_e, _dragObject] = getArgs(event, dragObject);
        const mouseCoord = getMouseCoord(_dragObject.domEvent, direction);

        const thumbPosition = getThumbPosition(
            draggingData.size,
            draggingData.offset,
            mouseCoord,
            thumbSize,
            draggingData.dragPointOffset
        );
        setThumbPosition(thumbPosition);
        const position = getScrollCoordByThumb(
            draggingData.size,
            thumbSize,
            contentSize,
            thumbPosition
        );
        props.onPositionchanged(undefined, position);
    }

    function wheelHandler(event: React.WheelEvent): void {
        let newPosition = position + calcWheelDelta(event.nativeEvent.deltaY);
        const maxPosition = contentSize - scrollbarSize;
        if (newPosition < 0) {
            newPosition = 0;
        } else if (newPosition > maxPosition) {
            newPosition = maxPosition;
        }
        props.onPositionchanged(undefined, newPosition);
        const newThumbPosition = getThumbCoordByScroll(
            scrollbarSize,
            thumbSize,
            contentSize,
            newPosition
        );
        setThumbPosition(newThumbPosition);
    }

    function getPaddingClasses(): string {
        const classes = [];
        if (direction === ScrollDirection.Vertical) {
            classes.push(
                `controls-padding_top-${paddings.start ? 'm' : '3xs'}`
            );
            classes.push(
                `controls-padding_bottom-${paddings.end ? 'm' : '3xs'}`
            );
        } else {
            classes.push(
                `controls-padding_left-${paddings.start ? 'm' : '3xs'}`
            );
            classes.push(
                `controls-padding_right-${paddings.end ? 'm' : '3xs'}`
            );
        }
        return classes.join(' ');
    }

    function getThumbWrapperStyle(): object {
        return {
            [direction === 'vertical' ? 'height' : 'width']: thumbSize + 'px',
            transform:
                `translate${direction === 'vertical' ? 'Y' : 'X'}(` +
                thumbPosition +
                'px)',
        };
    }

    function getClasses(): string {
        const classes = [];
        if (userAttrs && userAttrs.className) {
            classes.push(userAttrs.className);
        }
        classes.push(
            `controls_scroll_theme-${props.theme} controls-VScrollbar`
        );
        classes.push(
            `controls-VScrollbar_${direction} controls-VScrollbar_${direction}_size-${thumbThickness}`
        );
        classes.push(
            `controls-VScrollbar_horizontal-track_size-${thumbThickness}`
        );

        if (props.shouldSetMarginTop) {
            classes.push(
                `controls-VScrollbar_horizontal_margin-top_size-${thumbThickness}`
            );
        }

        if (dragging) {
            classes.push(`controls-VScrollbar_${direction}_dragging`);
            classes.push(
                `controls-VScrollbar_${direction}_dragging_size-${thumbThickness}`
            );
        }

        if (!thumbSize) {
            classes.push('ws-invisible');
        }

        if (direction === 'horizontal' && props.trackVisible) {
            classes.push(
                'controls-VScrollbar_horizontal-track controls-VScrollbar_horizontal-track'
            );
        }

        return classes.join(' ');
    }

    function getDnDEvents(): object {
        return {
            'on:customdragstart': [
                () => {
                    scrollbarStartDragHandler();
                },
            ],
            'on:dragmove': [
                (e, dragObject) => {
                    scrollbarOnDragHandler(e, dragObject);
                },
            ],
            'on:documentdragend': [
                () => {
                    scrollbarEndDragHandler();
                },
            ],
        };
    }

    function setThumbRef(el: HTMLDivElement): void {
        thumbRef.current = el;
    }

    return (
        <div
            ref={(el) => {
                scrollbarRef.current = el;
                forwardedRef?.(el);
            }}
            className={getClasses()}
            style={{
                ...((userAttrs && userAttrs.style) || {}),
                ...(props.style || {}),
            }}
            data-qa={`VScrollbar_${direction}`}
            onMouseDown={scrollbarMouseDownHandler}
            onTouchStart={touchStartHandler}
            onWheel={wheelHandler}
        >
            {createElement(
                DnDContainer,
                {
                    ref: dndRef,
                    content: thumbNode,
                    setThumbRef,
                    direction,
                    getPaddingClasses,
                    getThumbWrapperStyle,
                    thumbMouseDownHandler,
                    thumbStyle,
                    thumbThickness,
                    dragging,
                },
                undefined,
                { ...getDnDEvents() }
            )}
        </div>
    );
}

export default React.forwardRef(ScrollbarReact);
