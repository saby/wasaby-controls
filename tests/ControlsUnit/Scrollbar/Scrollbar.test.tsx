/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import ScrollbarReact from 'Controls/_scrollbar/ScrollbarReact';
import { WasabyEvents } from 'UICore/Events';

describe('Controls/_scrollbar/Scrollbar', () => {
    let container = null;
    beforeEach(() => {
        container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        WasabyEvents.destroyInstance(container);
        container.remove();
        container = null;
    });

    let resizeFuncMock: Function;
    class ResizeObserver {
        resizeFunc: Function;

        constructor(resizeCallback: Function) {
            this.resizeFunc = resizeCallback;
            resizeFuncMock = resizeCallback;
        }

        observe(): void {
            // mock
        }
        unobserve(): void {
            // mock
        }
        disconnect(): void {
            // mock
        }
    }

    it('Скроллбар строится с нулевой позицией', () => {
        act(() => {
            render(
                <ScrollbarReact
                    attrs={{}}
                    ref={jest.fn()}
                    contentSize={1000}
                    paddings={{ start: false, end: false }}
                ></ScrollbarReact>,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('Горизонтальный скроллбар строится с нулевой позицией', () => {
        act(() => {
            render(
                <ScrollbarReact
                    attrs={{}}
                    ref={jest.fn()}
                    contentSize={1000}
                    direction={'horizontal'}
                    paddings={{ start: false, end: false }}
                ></ScrollbarReact>,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('Скроллбар строится в ненулевой позицией с заданной опцией position (position = 10)', () => {
        const srcResizeObserver = window.ResizeObserver;
        window.ResizeObserver = ResizeObserver;

        act(() => {
            render(
                <ScrollbarReact
                    attrs={{}}
                    ref={jest.fn()}
                    position={10}
                    contentSize={1000}
                    paddings={{ start: false, end: false }}
                ></ScrollbarReact>,
                container
            );
            resizeFuncMock([
                {
                    contentRect: { height: 600 },
                },
            ]);
        });
        window.ResizeObserver = srcResizeObserver;
        expect(container).toMatchSnapshot();
    });

    it('Скроллбар строится с нижним отступом', () => {
        act(() => {
            render(
                <ScrollbarReact
                    attrs={{}}
                    ref={jest.fn()}
                    contentSize={1000}
                    paddings={{ start: false, end: true }}
                ></ScrollbarReact>,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });

    it('Скроллбар строится с опциями thumbThickness=s и thumbStyle=unaccented', () => {
        act(() => {
            render(
                <ScrollbarReact
                    attrs={{}}
                    ref={jest.fn()}
                    contentSize={1000}
                    thumbThickness={'s'}
                    thumbStyle={'unaccented'}
                    paddings={{ start: false, end: false }}
                ></ScrollbarReact>,
                container
            );
        });
        expect(container).toMatchSnapshot();
    });
});
