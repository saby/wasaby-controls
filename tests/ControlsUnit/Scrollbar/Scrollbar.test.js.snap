// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls/_scrollbar/Scrollbar Горизонтальный скроллбар строится с нулевой позицией 1`] = `
<div>
  <div
    class="controls_scroll_theme-undefined controls-VScrollbar controls-VScrollbar_horizontal controls-VScrollbar_horizontal_size-l controls-VScrollbar_horizontal-track_size-l ws-invisible"
    data-qa="VScrollbar_horizontal"
  >
    <div
      class="controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-l controls-VScrollbar__thumbWrapper_horizontal controls-padding_left-3xs controls-padding_right-3xs"
      style="transform: translateX(0px);"
      tabindex="0"
    >
      <div
        class="controls-VScrollbar__thumb controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_size-l controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_accented_horizontal controls-VScrollbar__thumb_horizontal controls-VScrollbar__thumb_horizontal_size-l "
        data-qa="VScrollbar__thumb"
      />
    </div>
  </div>
</div>
`;

exports[`Controls/_scrollbar/Scrollbar Скроллбар строится в ненулевой позицией с заданной опцией position (position = 10) 1`] = `
<div>
  <div
    class="controls_scroll_theme-undefined controls-VScrollbar controls-VScrollbar_vertical controls-VScrollbar_vertical_size-l controls-VScrollbar_horizontal-track_size-l"
    data-qa="VScrollbar_vertical"
  >
    <div
      class="controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-l controls-VScrollbar__thumbWrapper_vertical controls-padding_top-3xs controls-padding_bottom-3xs"
      style="transform: translateY(6px); height: 360px;"
      tabindex="0"
    >
      <div
        class="controls-VScrollbar__thumb controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_size-l controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_accented_vertical controls-VScrollbar__thumb_vertical controls-VScrollbar__thumb_vertical_size-l "
        data-qa="VScrollbar__thumb"
      />
    </div>
  </div>
</div>
`;

exports[`Controls/_scrollbar/Scrollbar Скроллбар строится с нижним отступом 1`] = `
<div>
  <div
    class="controls_scroll_theme-undefined controls-VScrollbar controls-VScrollbar_vertical controls-VScrollbar_vertical_size-l controls-VScrollbar_horizontal-track_size-l ws-invisible"
    data-qa="VScrollbar_vertical"
  >
    <div
      class="controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-l controls-VScrollbar__thumbWrapper_vertical controls-padding_top-3xs controls-padding_bottom-m"
      style="transform: translateY(0px);"
      tabindex="0"
    >
      <div
        class="controls-VScrollbar__thumb controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_size-l controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_accented_vertical controls-VScrollbar__thumb_vertical controls-VScrollbar__thumb_vertical_size-l "
        data-qa="VScrollbar__thumb"
      />
    </div>
  </div>
</div>
`;

exports[`Controls/_scrollbar/Scrollbar Скроллбар строится с нулевой позицией 1`] = `
<div>
  <div
    class="controls_scroll_theme-undefined controls-VScrollbar controls-VScrollbar_vertical controls-VScrollbar_vertical_size-l controls-VScrollbar_horizontal-track_size-l ws-invisible"
    data-qa="VScrollbar_vertical"
  >
    <div
      class="controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-l controls-VScrollbar__thumbWrapper_vertical controls-padding_top-3xs controls-padding_bottom-3xs"
      style="transform: translateY(0px);"
      tabindex="0"
    >
      <div
        class="controls-VScrollbar__thumb controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_size-l controls-VScrollbar__thumb_accented controls-VScrollbar__thumb_accented_vertical controls-VScrollbar__thumb_vertical controls-VScrollbar__thumb_vertical_size-l "
        data-qa="VScrollbar__thumb"
      />
    </div>
  </div>
</div>
`;

exports[`Controls/_scrollbar/Scrollbar Скроллбар строится с опциями thumbThickness=s и thumbStyle=unaccented 1`] = `
<div>
  <div
    class="controls_scroll_theme-undefined controls-VScrollbar controls-VScrollbar_vertical controls-VScrollbar_vertical_size-s controls-VScrollbar_horizontal-track_size-s ws-invisible"
    data-qa="VScrollbar_vertical"
  >
    <div
      class="controls-VScrollbar__thumbWrapper controls-VScrollbar__thumbWrapper_size-s controls-VScrollbar__thumbWrapper_vertical controls-padding_top-3xs controls-padding_bottom-3xs"
      style="transform: translateY(0px);"
      tabindex="0"
    >
      <div
        class="controls-VScrollbar__thumb controls-VScrollbar__thumb_unaccented controls-VScrollbar__thumb_size-s controls-VScrollbar__thumb_unaccented controls-VScrollbar__thumb_unaccented_vertical controls-VScrollbar__thumb_vertical controls-VScrollbar__thumb_vertical_size-s "
        data-qa="VScrollbar__thumb"
      />
    </div>
  </div>
</div>
`;
